### PR TITLE
Attempt to get docker image building on a GDAL 3.9 base.

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Reason for this pull request

The latest osgeo/gdal/ubuntu base images use Ubuntu 24.04 with Python 3.12 and GDAL 3.9

This requires a pretty major uplift of our test docker image build.

Edit: That was a nightmare, but seeing as I actually got working we will run with it at least in the short-term.  In the longer term a complete rewrite with something simpler and more maintainable is being discussed.

### Proposed changes

- Added minimum versions for a number of packages that have only recently become Python 3.12 compatible.
- Manually build (`--no-binary`) wheels for packages which need to link to GDAL or otherwise do not have 3.12 compatible wheels available on PyPI.

 - [ ] Tests added / passed

----
📚 Documentation preview 📚: https://datacube-core--1587.org.readthedocs.build/en/1587/

### CODE REVIEW

File: docker/nobinary.txt [hunk 1]
```diff
@@ -1 +1,5 @@
-# Empty for now
+# libraries to compile with local gdal
+--no-binary rasterio
+--no-binary fiona
+--no-binary shapely
+--no-binary cf-units
```
Comment (Kirill888) [inline docker/nobinary.txt:5]: `pyproj`?

Comment (SpacemanPaul) [reply docker/nobinary.txt:5]: Pyproj seems to be installing fine from pip without `--no-binary` - I only added them if the binary wheel didn't work.

Comment (Kirill888) [reply docker/nobinary.txt:5]: my suggestion is don't bother with non-conda installs, but this looks like your gdal is missing netcdf support OR your build stage has more libs than your use stage.

conda envs are somewhat larger than pip, but at least you don't have that nightmare to deal with. Alternatively go all out and build EVERYTHING from source and stop mixing pre-compiled things from pip with locally compiled code.

File: datacube/utils/geometry/_base.py [hunk 2]
```diff
@@ -853,9 +853,6 @@ def multipoint(coords: CoordList, crs: MaybeCRS) -> Geometry:
     """
     Create a 2D MultiPoint Geometry
 
-    >>> multipoint([(10, 10), (20, 20)], None)
-    Geometry(MULTIPOINT (10 10, 20 20), None)
-
```
Comment (SpacemanPaul) [inline datacube/utils/geometry/_base.py:858]: This test fails - looks like shapely has changed the str representation of multipoint geometries.

File: setup.py [hunk 3]
```diff
@@ -21,6 +21,7 @@
     'beautifulsoup4',
     'nbsphinx',
     'pydata-sphinx-theme==0.9.0',
+    'twine'
```
Comment (omad) [inline setup.py]: Why do the docs require twine!?

File: docker/nobinary.txt (hunk 1 above)
Comment (SpacemanPaul) [reply docker/nobinary.txt:5]: Now I've got it working we'll probably stick with this for the short term.  In the longer term, @omad has ideas for a long overdue complete restructure the way tests are run in github (including dropping docker all together).  An Issue will be created soon.

**Review by omad — APPROVED** [review]
Thanks Paul, so much hurt, glad it's working again for now.

**Pull request merged**